src/traffic_server/InkAPI.cc

```diff
@@ -1873,6 +1887,122 @@ _TSfree(void *ptr)
   ats_free(ptr);
 }
 
+// Support code for TSHeapBuf.
+namespace
+{
+class HeapBuf
```

Comment:
That seems like it would require three memory allocations instead of one.  This is meant to be a general purpose type, so pessimizing it could have a significant performance impact, depending how it's used in the future.

Reply:
Should only have to heap allocate the string as you do with `_S` here.

Reply:
TSHeapBuf would have to be of type `std::unique_ptr<std::string> *`.  The `std::uniue_ptr<std::string>` instance would be dynamically allocated (allocation 1).  It would point to a `std::string` instance in the heap (allocation 2).  The `std::string` instance would allocate a buffer for the data (allocation 3).

Reply:
I meant that effectively, HeapBuf could be written as `using HeapBuf = std::unique_ptr<std::string>` and TSHeapBuf would be (opaquely) `*std::string`.

Reply:
OK I see what you mean now.  But that will still mean doubling the number of memory allocations.  Also mice that HeapBuf encapsulates the reinterpret_casts.